### DISCUSSION TITLE
NPM packages in pb_hooks

### DISCUSSION THREAD

**Discussion by charbs-io** [pocketbase/pocketbase#3175] [Q&A]
Has anyone figured out a feasible way to `require` npm packages in `.pb.js` files in the `pb_hooks` directory?

I understand the limitations stated [here](https://pocketbase.io/docs/js-overview/#relative-paths), but even with a simple example I'm still having trouble at making it work...

As a simple test:
I have a `package.json` file and my `node_modules` directory at the root of the pocketbase directory. 

Then I have a `pb_hooks/config/uuid.js`
```
const { v4: uuidv4 } = require("../../node_modules/uuid/dist/commonjs-browser/index.js");

module.exports = {
  uuid: () => {
    return uuidv4();
  },
};
```

And then a `uuid.pb.js`
```
routerAdd("POST", "/api/extend/uuid", (c) => {
  const uuid = require(`${__hooks}/config/uuid.js`);

  return c.json(200, {
    message: uuid(),
  });
});
```

I'm getting a whole of errors depending on what I change/do.

**Comment by ganigeorgiev** [comment]
What are the errors you are getting?

Note that the relative path is relative to the executable/current working dir and not to the pb_hooks dir (see https://pocketbase.io/docs/js-overview/#relative-paths).

**Comment by charbs-io** [comment]
When I fix some silly errors (on my end!), I get `Error: crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported at rng (/Users/charbelayoub/Sites/moneymapr/pocketbase/node_modules/uuid/dist/commonjs-browser/rng.js:20:13(23))`

So I guess that's me hitting the limitations 😭

**Reply by charbs-io** [reply]
In this case (not that I would need the uuid library) I would need to add some polyfills I guess?

**Comment by charbs-io** [comment]
The most useful usecase I'm testing is being able to import and work with the Stripe API...

**Reply by ganigeorgiev** [reply]
Unfortunately I don't think the Stripe JS library can be loaded in `pb_hooks` and you'll have to interact directly with their [web API](https://stripe.com/docs/api) via `$http.send()`.

**Reply by ganigeorgiev** [reply]
I guess a common operation would be verifying a webhook event, but for that I think you'll need to calculate a sha256 hash of the payload, which may not be easy to do at the moment.

With the next release I'll consider exposing some common Go crypto primitives under the [`$security` namespace](https://pocketbase.io/jsvm/modules/_security.html).

**Reply by charbs-io** [reply]
@ganigeorgiev - just tested the `$http.send()` method - works okay actually - hadn't considered it. 

Yes i think that would be useful actually...
